### PR TITLE
refactor: improve resource plugin onboarding

### PR DESCRIPTION
**Pull request by jayzhang**
1. update Plugin interface: 
image
2. adopt `typedi` as IOC container to manage all resource plugins
Here is the single truth of resource plugin identifiers:
image

3. adjust resource plugins's code and UTs accordingly:
- add `name` and `displayName`
- implement activate API
- add decorator for resource plugin class
image

### CODE REVIEW

**Review by 1yefuwang1 — APPROVED** [review]

**Review by KennethBWSong — APPROVED** [review]
Approve for aad and simple auth.

**Review by xzf0587 — APPROVED** [review]

**Review by nliu-ms — APPROVED** [review]
Approved for app studio plugin

**Review by swatDong — APPROVED** [review]
Local debug plugin looks good to me.

**Review by jiayi-ruan — APPROVED** [review]
Approve for spfx

File: packages/fx-core/src/plugins/resource/function/index.ts [hunk 1]
```diff
@@ -26,9 +27,31 @@ import { FxResult, FunctionPluginResultFactory as ResultFactory } from "./result
...
+} from "../../solution/fx-solution/question";
+import { Service } from "typedi";
+import { ResourcePlugins } from "../../solution/fx-solution/ResourcePluginContainer";
+@Service(ResourcePlugins.FunctionPlugin)
 export class FunctionPlugin implements Plugin {
+  name = "fx-resource-function";
+  displayName = "Azure Function";
+  activate(solutionSettings: AzureSolutionSettings): boolean {
+    const cap = solutionSettings.capabilities || [];
+    const azureResources = solutionSettings.azureResources ? solutionSettings.azureResources : [];
```
Comment (hund030) [inline packages/fx-core/src/plugins/resource/function/index.ts:45]: NIT: solutionSettings.azureResources ?? []

**Review by hund030 — APPROVED** [review]

**Review by XiaofuHuang — APPROVED** [review]

**Review by IvanJobs — APPROVED** [review]
lgtm

**Review by YitongFeng-git — APPROVED** [review]

**Pull request merged**